File: integration-tests/TestAPI-bindings/kotlin/src/test/kotlin/com/cricut/testapi/PrimitiveTests.kt
package com.cricut.testapi

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

internal class PrimitiveTests {
    @Test
    fun testPrimitiveValues() {
        assertEquals(Primitives.falseBool, false)
        assertEquals(Primitives.trueBool, true)
        assertEquals(Primitives.zeroUInt8, 0.toUByte())
        assertEquals(Primitives.minUInt8, UByte.MIN_VALUE)
        assertEquals(Primitives.maxUInt8, UByte.MAX_VALUE)
        assertEquals(Primitives.zeroUInt16, 0.toUShort())
        assertEquals(Primitives.minUInt16, UShort.MIN_VALUE)
        assertEquals(Primitives.maxUInt16, UShort.MAX_VALUE)
        assertEquals(Primitives.zeroUInt32, 0.toUInt())
        assertEquals(Primitives.minUInt32, UInt.MIN_VALUE)
        assertEquals(Primitives.maxUInt32, UInt.MAX_VALUE)
        // TODO: UInt64
//        assertEquals(Primitives.zeroUInt64, 0.toULong())
//        assertEquals(Primitives.minUInt64, ULong.MIN_VALUE)
//        assertEquals(Primitives.maxUInt64, ULong.MAX_VALUE)
        assertEquals(Primitives.zeroInt8, 0.toByte())
        assertEquals(Primitives.minInt8, Byte.MIN_VALUE)
        assertEquals(Primitives.maxInt8, Byte.MAX_VALUE)
        assertEquals(Primitives.zeroInt16, 0.toShort())
        assertEquals(Primitives.minInt16, Short.MIN_VALUE)
        assertEquals(Primitives.maxInt16, Short.MAX_VALUE)
        assertEquals(Primitives.zeroInt32, 0.toInt())
        assertEquals(Primitives.minInt32, Int.MIN_VALUE)
        assertEquals(Primitives.maxInt32, Int.MAX_VALUE)
        // TODO: Int64
//        assertEquals(Primitives.zeroInt64, 0.toLong())
//        assertEquals(Primitives.minInt64, Long.MIN_VALUE)
//        assertEquals(Primitives.maxInt64, Long.MAX_VALUE)
        assertEquals(Primitives.zeroFloat, 0.0f)
        assertEquals(Primitives.minFloat, -Float.MAX_VALUE)
        assertEquals(Primitives.maxFloat, Float.MAX_VALUE)
        assertEquals(Primitives.zeroDouble, 0.0)
        assertEquals(Primitives.minDouble, -Double.MAX_VALUE)
        assertEquals(Primitives.maxDouble, Double.MAX_VALUE)
    }

    @Test
    fun testArraysOfPrimitiveValues() {
        assert(Primitives.manyBool.zip(arrayOf(false, true)).all { it.first == it.second })
        assert(Primitives.manyUInt8.zip(arrayOf(0.toUByte(), UByte.MIN_VALUE, UByte.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyUInt16.zip(arrayOf(0.toUShort(), UShort.MIN_VALUE, UShort.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyUInt32.zip(arrayOf(0.toUInt(), UInt.MIN_VALUE, UInt.MAX_VALUE)).all { it.first == it.second })
        // TODO: UInt64
//        assert(Primitives.manyUInt64.zip(arrayOf(0.toULong(), ULong.MIN_VALUE, ULong.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyInt8.zip(arrayOf(0.toByte(), Byte.MIN_VALUE, Byte.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyInt16.zip(arrayOf(0.toShort(), Short.MIN_VALUE, Short.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyInt32.zip(arrayOf(0.toInt(), Int.MIN_VALUE, Int.MAX_VALUE)).all { it.first == it.second })
        // TODO: Int64
//        assert(Primitives.manyInt64.zip(arrayOf(0.toLong(), Long.MIN_VALUE, Long.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyFloat.zip(arrayOf(0.toFloat(), -Float.MAX_VALUE, Float.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyDouble.zip(arrayOf(0.toDouble(), -Double.MAX_VALUE, Double.MAX_VALUE)).all { it.first == it.second })
    }

    @Test
    fun testArraysOfOptionalPrimitiveValues() {
        assert(Primitives.manyMaybeBool.zip(arrayOf(null, false, true)).all { it.first == it.second })
        assert(Primitives.manyMaybeUInt8.zip(arrayOf(null, 0.toUByte(), UByte.MIN_VALUE, UByte.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyMaybeUInt16.zip(arrayOf(null, 0.toUShort(), UShort.MIN_VALUE, UShort.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyMaybeUInt32.zip(arrayOf(null, 0.toUInt(), UInt.MIN_VALUE, UInt.MAX_VALUE)).all { it.first == it.second })
        // TODO: UInt64
//        assert(Primitives.manyMaybeUInt64.zip(arrayOf(null, 0.toULong(), ULong.MIN_VALUE, ULong.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyMaybeInt8.zip(arrayOf(null, 0.toByte(), Byte.MIN_VALUE, Byte.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyMaybeInt16.zip(arrayOf(null, 0.toShort(), Short.MIN_VALUE, Short.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyMaybeInt32.zip(arrayOf(null, 0.toInt(), Int.MIN_VALUE, Int.MAX_VALUE)).all { it.first == it.second })
        // TODO: Int64
//        assert(Primitives.manyMaybeInt64.zip(arrayOf(null, 0.toLong(), Long.MIN_VALUE, Long.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyMaybeFloat.zip(arrayOf(null, 0.toFloat(), -Float.MAX_VALUE, Float.MAX_VALUE)).all { it.first == it.second })
        assert(Primitives.manyMaybeDouble.zip(arrayOf(null, 0.toDouble(), -Double.MAX_VALUE, Double.MAX_VALUE)).all { it.first == it.second })
    }

    @Test
    fun testFunctionsTakingAndReturningPrimitiveTypes() {
        assertEquals(Primitives.echoBool(false), false)
        assertEquals(Primitives.echoBool(true), true)
        assertEquals(Primitives.echoUInt8(0.toUByte()), 0.toUByte())
        assertEquals(Primitives.echoUInt8(UByte.MIN_VALUE), UByte.MIN_VALUE)
        assertEquals(Primitives.echoUInt8(UByte.MAX_VALUE), UByte.MAX_VALUE)
        assertEquals(Primitives.echoUInt16(0.toUShort()), 0.toUShort())
        assertEquals(Primitives.echoUInt16(UShort.MIN_VALUE), UShort.MIN_VALUE)
        assertEquals(Primitives.echoUInt16(UShort.MAX_VALUE), UShort.MAX_VALUE)
        assertEquals(Primitives.echoUInt32(0.toUInt()), 0.toUInt())
        assertEquals(Primitives.echoUInt32(UInt.MIN_VALUE), UInt.MIN_VALUE)
        assertEquals(Primitives.echoUInt32(UInt.MAX_VALUE), UInt.MAX_VALUE)
        // TODO: UInt64
//        assertEquals(Primitives.echoUInt64(0.toULong()), 0.toULong())
//        assertEquals(Primitives.echoUInt64(ULong.MIN_VALUE), ULong.MIN_VALUE)
//        assertEquals(Primitives.echoUInt64(ULong()), ULong())
        assertEquals(Primitives.echoInt8(0.toByte()), 0.toByte())
        assertEquals(Primitives.echoInt8(Byte.MIN_VALUE), Byte.MIN_VALUE)
        assertEquals(Primitives.echoInt8(Byte.MAX_VALUE), Byte.MAX_VALUE)
        assertEquals(Primitives.echoInt16(0.toShort()), 0.toShort())
        assertEquals(Primitives.echoInt16(Short.MIN_VALUE), Short.MIN_VALUE)
        assertEquals(Primitives.echoInt16(Short.MAX_VALUE), Short.MAX_VALUE)
        assertEquals(Primitives.echoInt32(0.toInt()), 0.toInt())
        assertEquals(Primitives.echoInt32(Int.MIN_VALUE), Int.MIN_VALUE)
        assertEquals(Primitives.echoInt32(Int.MAX_VALUE), Int.MAX_VALUE)
        // TODO: Int64
//        assertEquals(Primitives.echoInt64(null), null)
//        assertEquals(Primitives.echoInt64(0.toLong()), 0.toLong())
//        assertEquals(Primitives.echoInt64(Long.MIN_VALUE), Long.MIN_VALUE)
//        assertEquals(Primitives.echoInt64(Long.MAX_VALUE), Long.MAX_VALUE)
        assertEquals(Primitives.echoFloat(0.0f), 0.0f)
        assertEquals(Primitives.echoFloat(-Float.MAX_VALUE), -Float.MAX_VALUE)
        assertEquals(Primitives.echoFloat(Float.MAX_VALUE), Float.MAX_VALUE)
        assertEquals(Primitives.echoDouble(0.0), 0.0)
        assertEquals(Primitives.echoDouble(-Double.MAX_VALUE), -Double.MAX_VALUE)
        assertEquals(Primitives.echoDouble(Double.MAX_VALUE), Double.MAX_VALUE)
    }

    @Test
    fun testFunctionsTakingReturningOptionalPrimitiveTypes() {
        assertEquals(Primitives.maybeEchoBool(null), null)
        assertEquals(Primitives.maybeEchoBool(false), false)
        assertEquals(Primitives.maybeEchoBool(true), true)
        assertEquals(Primitives.maybeEchoUInt8(null), null)
        assertEquals(Primitives.maybeEchoUInt8(0.toUByte()), 0.toUByte())
        assertEquals(Primitives.maybeEchoUInt8(UByte.MIN_VALUE), UByte.MIN_VALUE)
        assertEquals(Primitives.maybeEchoUInt8(UByte.MAX_VALUE), UByte.MAX_VALUE)
        assertEquals(Primitives.maybeEchoUInt16(null), null)
        assertEquals(Primitives.maybeEchoUInt16(0.toUShort()), 0.toUShort())
        assertEquals(Primitives.maybeEchoUInt16(UShort.MIN_VALUE), UShort.MIN_VALUE)
        assertEquals(Primitives.maybeEchoUInt16(UShort.MAX_VALUE), UShort.MAX_VALUE)
        assertEquals(Primitives.maybeEchoUInt32(null), null)
        assertEquals(Primitives.maybeEchoUInt32(0.toUInt()), 0.toUInt())
        assertEquals(Primitives.maybeEchoUInt32(UInt.MIN_VALUE), UInt.MIN_VALUE)
        assertEquals(Primitives.maybeEchoUInt32(UInt.MAX_VALUE), UInt.MAX_VALUE)
        // TODO: UInt64
//        assertEquals(Primitives.maybeEchoUInt64(null), null)
//        assertEquals(Primitives.maybeEchoUInt64(0.toULong()), 0.toULong())
//        assertEquals(Primitives.maybeEchoUInt64(ULong.MIN_VALUE), ULong.MIN_VALUE)
//        assertEquals(Primitives.maybeEchoUInt64(ULong()), ULong())
        assertEquals(Primitives.maybeEchoInt8(null), null)
        assertEquals(Primitives.maybeEchoInt8(0.toByte()), 0.toByte())
        assertEquals(Primitives.maybeEchoInt8(Byte.MIN_VALUE), Byte.MIN_VALUE)
        assertEquals(Primitives.maybeEchoInt8(Byte.MAX_VALUE), Byte.MAX_VALUE)
        assertEquals(Primitives.maybeEchoInt16(null), null)
        assertEquals(Primitives.maybeEchoInt16(0.toShort()), 0.toShort())
        assertEquals(Primitives.maybeEchoInt16(Short.MIN_VALUE), Short.MIN_VALUE)
        assertEquals(Primitives.maybeEchoInt16(Short.MAX_VALUE), Short.MAX_VALUE)
        assertEquals(Primitives.maybeEchoInt32(null), null)
        assertEquals(Primitives.maybeEchoInt32(0.toInt()), 0.toInt())
        assertEquals(Primitives.maybeEchoInt32(Int.MIN_VALUE), Int.MIN_VALUE)
        assertEquals(Primitives.maybeEchoInt32(Int.MAX_VALUE), Int.MAX_VALUE)
        // TODO: Int64
//        assertEquals(Primitives.maybeEchoInt64(null), null)
//        assertEquals(Primitives.maybeEchoInt64(0.toLong()), 0.toLong())
//        assertEquals(Primitives.maybeEchoInt64(Long.MIN_VALUE), Long.MIN_VALUE)
//        assertEquals(Primitives.maybeEchoInt64(Long.MAX_VALUE), Long.MAX_VALUE)
        assertEquals(Primitives.maybeEchoFloat(null), null)
        assertEquals(Primitives.maybeEchoFloat(0.0f), 0.0f)
        assertEquals(Primitives.maybeEchoFloat(-Float.MAX_VALUE), -Float.MAX_VALUE)
        assertEquals(Primitives.maybeEchoFloat(Float.MAX_VALUE), Float.MAX_VALUE)
        assertEquals(Primitives.maybeEchoDouble(null), null)
        assertEquals(Primitives.maybeEchoDouble(0.0), 0.0)
        assertEquals(Primitives.maybeEchoDouble(-Double.MAX_VALUE), -Double.MAX_VALUE)
        assertEquals(Primitives.maybeEchoDouble(Double.MAX_VALUE), Double.MAX_VALUE)
    }
}

File: integration-tests/TestAPI-bindings/c-sharp/TestAPI.Tests/PrimitiveTests.cs
using System;
using Xunit;
using Cricut.TestAPI;

namespace TestAPI.Tests {
    public class PrimitiveTests {
        [Fact]
        void testPrimitiveValues() {
            Assert.False(Primitives.falseBool);
            Assert.True(Primitives.trueBool);
            Assert.Equal(0u, Primitives.zeroUInt8);
            Assert.Equal(byte.MinValue, Primitives.minUInt8);
            Assert.Equal(byte.MaxValue, Primitives.maxUInt8);
            Assert.Equal(0u, Primitives.zeroUInt16);
            Assert.Equal(ushort.MinValue, Primitives.minUInt16);
            Assert.Equal(ushort.MaxValue, Primitives.maxUInt16);
            Assert.Equal(0u, Primitives.zeroUInt32);
            Assert.Equal(uint.MinValue, Primitives.minUInt32);
            Assert.Equal(uint.MaxValue, Primitives.maxUInt32);
            Assert.Equal(0u, Primitives.zeroUInt64);
            Assert.Equal(ulong.MinValue, Primitives.minUInt64);
            Assert.Equal(ulong.MaxValue, Primitives.maxUInt64);
            Assert.Equal(0, Primitives.zeroInt8);
            Assert.Equal(sbyte.MinValue, Primitives.minInt8);
            Assert.Equal(sbyte.MaxValue, Primitives.maxInt8);
            Assert.Equal(0, Primitives.zeroInt16);
            Assert.Equal(short.MinValue, Primitives.minInt16);
            Assert.Equal(short.MaxValue, Primitives.maxInt16);
            Assert.Equal(0, Primitives.zeroInt32);
            Assert.Equal(int.MinValue, Primitives.minInt32);
            Assert.Equal(int.MaxValue, Primitives.maxInt32);
            Assert.Equal(0, Primitives.zeroInt64);
            Assert.Equal(long.MinValue, Primitives.minInt64);
            Assert.Equal(long.MaxValue, Primitives.maxInt64);
            Assert.Equal(0.0f, Primitives.zeroFloat);
            Assert.Equal(-float.MaxValue, Primitives.minFloat);
            Assert.Equal(float.MaxValue, Primitives.maxFloat);
            Assert.Equal(0.0, Primitives.zeroDouble);
            Assert.Equal(-double.MaxValue, Primitives.minDouble);
            Assert.Equal(double.MaxValue, Primitives.maxDouble);
        }

        [Fact]
        void testArraysOfPrimitiveValues() {
            Assert.Equal(new bool[] { false, true }, Primitives.manyBool);
            Assert.Equal(new byte[] { (byte)0, byte.MinValue, byte.MaxValue }, Primitives.manyUInt8);
            Assert.Equal(new ushort[] { (ushort)0, ushort.MinValue, ushort.MaxValue }, Primitives.manyUInt16);
            Assert.Equal(new uint[] { 0u, uint.MinValue, uint.MaxValue }, Primitives.manyUInt32);
            Assert.Equal(new ulong[] { 0u, ulong.MinValue, ulong.MaxValue }, Primitives.manyUInt64);
            Assert.Equal(new sbyte[] { 0, sbyte.MinValue, sbyte.MaxValue }, Primitives.manyInt8);
            Assert.Equal(new short[] { 0, short.MinValue, short.MaxValue }, Primitives.manyInt16);
            Assert.Equal(new int[] { 0, int.MinValue, int.MaxValue }, Primitives.manyInt32);
            Assert.Equal(new long[] { 0, long.MinValue, long.MaxValue }, Primitives.manyInt64);
            Assert.Equal(new float[] { 0.0f, -float.MaxValue, float.MaxValue }, Primitives.manyFloat);
            Assert.Equal(new double[] { 0.0, -double.MaxValue, double.MaxValue }, Primitives.manyDouble);
        }

        [Fact]
        void testArraysOfOptionalPrimitiveValues() {
            Assert.Equal(new bool?[] { null, false, true }, Primitives.manyMaybeBool);
            Assert.Equal(new byte?[] { null, (byte)0, byte.MinValue, byte.MaxValue }, Primitives.manyMaybeUInt8);
            Assert.Equal(new ushort?[] { null, 0, ushort.MinValue, ushort.MaxValue }, Primitives.manyMaybeUInt16);
            Assert.Equal(new uint?[] { null, 0, uint.MinValue, uint.MaxValue }, Primitives.manyMaybeUInt32);
            Assert.Equal(new ulong?[] { null, 0, ulong.MinValue, ulong.MaxValue }, Primitives.manyMaybeUInt64);
            Assert.Equal(new sbyte?[] { null, 0, sbyte.MinValue, sbyte.MaxValue }, Primitives.manyMaybeInt8);
            Assert.Equal(new short?[] { null, 0, short.MinValue, short.MaxValue }, Primitives.manyMaybeInt16);
            Assert.Equal(new int?[] { null, 0, int.MinValue, int.MaxValue }, Primitives.manyMaybeInt32);
            Assert.Equal(new long?[] { null, 0, long.MinValue, long.MaxValue }, Primitives.manyMaybeInt64);
            Assert.Equal(new float?[] { null, 0.0f, -float.MaxValue, float.MaxValue }, Primitives.manyMaybeFloat);
            Assert.Equal(new double?[] { null, 0.0, -double.MaxValue, double.MaxValue }, Primitives.manyMaybeDouble);
        }

        [Fact]
        void testVoidctionsTakingAndReturningPrimitiveTypes() {
            Assert.False(Primitives.echoBool(false));
            Assert.True(Primitives.echoBool(true));
            Assert.Equal(0, Primitives.echoUInt8(0));
            Assert.Equal(byte.MinValue, Primitives.echoUInt8(byte.MinValue));
            Assert.Equal(byte.MaxValue, Primitives.echoUInt8(byte.MaxValue));
            Assert.Equal(0, Primitives.echoUInt16(0));
            Assert.Equal(ushort.MinValue, Primitives.echoUInt16(ushort.MinValue));
            Assert.Equal(ushort.MaxValue, Primitives.echoUInt16(ushort.MaxValue));
            Assert.Equal(0u, Primitives.echoUInt32(0u));
            Assert.Equal(uint.MinValue, Primitives.echoUInt32(uint.MinValue));
            Assert.Equal(uint.MaxValue, Primitives.echoUInt32(uint.MaxValue));
            Assert.Equal(0u, Primitives.echoUInt64(0u));
            Assert.Equal(ulong.MinValue, Primitives.echoUInt64(ulong.MinValue));
            Assert.Equal(ulong.MaxValue, Primitives.echoUInt64(ulong.MaxValue));
            Assert.Equal(0, Primitives.echoInt8(0));
            Assert.Equal(sbyte.MinValue, Primitives.echoInt8(sbyte.MinValue));
            Assert.Equal(sbyte.MaxValue, Primitives.echoInt8(sbyte.MaxValue));
            Assert.Equal(0, Primitives.echoInt16(0));
            Assert.Equal(short.MinValue, Primitives.echoInt16(short.MinValue));
            Assert.Equal(short.MaxValue, Primitives.echoInt16(short.MaxValue));
            Assert.Equal(0, Primitives.echoInt32(0));
            Assert.Equal(int.MinValue, Primitives.echoInt32(int.MinValue));
            Assert.Equal(int.MaxValue, Primitives.echoInt32(int.MaxValue));
            Assert.Equal(0, Primitives.echoInt64(0));
            Assert.Equal(long.MinValue, Primitives.echoInt64(long.MinValue));
            Assert.Equal(long.MaxValue, Primitives.echoInt64(long.MaxValue));
            Assert.Equal(0.0f, Primitives.echoFloat(0.0f));
            Assert.Equal(-float.MaxValue, Primitives.echoFloat(-float.MaxValue));
            Assert.Equal(float.MaxValue, Primitives.echoFloat(float.MaxValue));
            Assert.Equal(0.0, Primitives.echoDouble(0.0));
            Assert.Equal(-double.MaxValue, Primitives.echoDouble(-double.MaxValue));
            Assert.Equal(double.MaxValue, Primitives.echoDouble(double.MaxValue));
        }

        [Fact]
        void testVoidctionsTakingReturningOptionalPrimitiveTypes() {
            Assert.Null(Primitives.maybeEchoBool(null));
            Assert.False(Primitives.maybeEchoBool(false));
            Assert.True(Primitives.maybeEchoBool(true));
            Assert.Null(Primitives.maybeEchoUInt8(null));
            Assert.Equal((byte) 0, Primitives.maybeEchoUInt8((byte)0));
            Assert.Equal(byte.MinValue, Primitives.maybeEchoUInt8(byte.MinValue));
            Assert.Equal(byte.MaxValue, Primitives.maybeEchoUInt8(byte.MaxValue));
            Assert.Null(Primitives.maybeEchoUInt16(null));
            Assert.Equal((ushort)0, Primitives.maybeEchoUInt16((ushort)0));
            Assert.Equal(ushort.MinValue, Primitives.maybeEchoUInt16(ushort.MinValue));
            Assert.Equal(ushort.MaxValue, Primitives.maybeEchoUInt16(ushort.MaxValue));
            Assert.Null(Primitives.maybeEchoUInt32(null));
            Assert.Equal(0u, Primitives.maybeEchoUInt32(0u));
            Assert.Equal(uint.MinValue, Primitives.maybeEchoUInt32(uint.MinValue));
            Assert.Equal(uint.MaxValue, Primitives.maybeEchoUInt32(uint.MaxValue));
            Assert.Null(Primitives.maybeEchoUInt64(null));
            Assert.Equal(0u, Primitives.maybeEchoUInt64(0u));
            Assert.Equal(ulong.MinValue, Primitives.maybeEchoUInt64(ulong.MinValue));
            Assert.Equal(ulong.MaxValue, Primitives.maybeEchoUInt64(ulong.MaxValue));
            Assert.Null(Primitives.maybeEchoInt8(null));
            Assert.Equal((sbyte)0, Primitives.maybeEchoInt8((sbyte)0));
            Assert.Equal(sbyte.MinValue, Primitives.maybeEchoInt8(sbyte.MinValue));
            Assert.Equal(sbyte.MaxValue, Primitives.maybeEchoInt8(sbyte.MaxValue));
            Assert.Null(Primitives.maybeEchoInt16(null));
            Assert.Equal((short)0, Primitives.maybeEchoInt16((short)0));
            Assert.Equal(short.MinValue, Primitives.maybeEchoInt16(short.MinValue));
            Assert.Equal(short.MaxValue, Primitives.maybeEchoInt16(short.MaxValue));
            Assert.Null(Primitives.maybeEchoInt32(null));
            Assert.Equal(0, Primitives.maybeEchoInt32(0));
            Assert.Equal(int.MinValue, Primitives.maybeEchoInt32(int.MinValue));
            Assert.Equal(int.MaxValue, Primitives.maybeEchoInt32(int.MaxValue));
            Assert.Null(Primitives.maybeEchoInt64(null));
            Assert.Equal(0, Primitives.maybeEchoInt64(0));
            Assert.Equal(long.MinValue, Primitives.maybeEchoInt64(long.MinValue));
            Assert.Equal(long.MaxValue, Primitives.maybeEchoInt64(long.MaxValue));
            Assert.Null(Primitives.maybeEchoFloat(null));
            Assert.Equal(0.0f, Primitives.maybeEchoFloat(0.0f));
            Assert.Equal(-float.MaxValue, Primitives.maybeEchoFloat(-float.MaxValue));
            Assert.Equal(float.MaxValue, Primitives.maybeEchoFloat(float.MaxValue));
            Assert.Null(Primitives.maybeEchoDouble(null));
            Assert.Equal(0.0, Primitives.maybeEchoDouble(0.0));
            Assert.Equal(-Double.MaxValue, Primitives.maybeEchoDouble(-Double.MaxValue));
            Assert.Equal(Double.MaxValue, Primitives.maybeEchoDouble(Double.MaxValue));
        }

        [Fact]
        void testVoidctionsTakingClosuresWithPrimitiveTypes() {
            // TODO: Fix breakage
//        Assert.Equal(Primitives.valueMapper(10) { (it ?: 0) * 2 }, 20);
        }

        [Fact]
        void testObjectsWithPrimitiveMembers() {
            Assert.Equal(Primitives.PrimitiveHolder.staticPropery, new byte?[] { null, (byte)0, byte.MinValue, byte.MaxValue });
            Assert.Equal(Primitives.PrimitiveHolder.staticMutablePropery, new byte?[] { null, (byte)0, byte.MinValue, byte.MaxValue });
            Primitives.PrimitiveHolder.staticMutablePropery = new byte?[] { (byte)100, (byte)200 };
            Assert.Equal(Primitives.PrimitiveHolder.staticMutablePropery, new byte?[] { (byte)100, (byte)200 });
            var s = Primitives.defaultPrimitiveHolder;
            Assert.Equal(s, Primitives.defaultPrimitiveHolder);
            s.b = !s.b;
            Assert.NotEqual(s, Primitives.defaultPrimitiveHolder);
        }
    }
}
